### PR TITLE
[pt2] add meta function for `_linalg_eigh`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #100965
* __->__ #100964

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -404,11 +404,33 @@ def checkInputsSolver(
...
-    uplo_uppercase = uplo.upper()
-    assert (
-        len(uplo) == 1 and uplo_uppercase == "U" or uplo_uppercase == "L"
-    ), f"Expected UPLO argument to be 'L' or 'U', but got {uplo}"
+def checkUplo(UPLO: str):
+    UPLO_uppercase = UPLO.upper()
+    check(
+        len(UPLO) == 1 and (UPLO_uppercase == "U" or UPLO_uppercase == "L"),
+        lambda: f"Expected UPLO argument to be 'L' or 'U', but got {UPLO}",
+    )
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:412]: Added parens after `and` to match the code in core. Capitalized `uplo` to match the yaml file param name.

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -404,11 +404,33 @@ def checkInputsSolver(
...
+        eigenvectors.as_strided_(
+            shape, make_contiguous_strides_for(shape, row_major=False)
+        )
+    else:
+        eigenvectors = A.new_empty([0])
+
+    shape.pop()
+    eigenvalues = A.new_empty(shape, dtype=toRealValueType(A.dtype))
+
+    return eigenvalues, eigenvectors
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:433]: See `TORCH_META_FUNC(_linalg_eigh)` in core.

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -404,11 +404,46 @@ def checkInputsSolver(
...
+    vals = A.new_empty(shape, dtype=toRealValueType(A.dtype))
+
+    if eigenvalues is not None and eigenvectors is not None:
+        assert isinstance(eigenvalues, TensorLike)
+        assert isinstance(eigenvectors, TensorLike)
+        eigenvalues = _maybe_resize_out(eigenvalues, vals.shape)
+        eigenvectors = _maybe_resize_out(eigenvectors, vecs.shape)
+        _safe_copy_out(copy_from=vals, copy_to=eigenvalues)  # type: ignore[arg-type]
+        _safe_copy_out(copy_from=vecs, copy_to=eigenvectors)  # type: ignore[arg-type]
+        return eigenvalues, eigenvectors
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:444]: `out_wrapper` didn't work with `TestMetaCPU.test_dispatch_meta_outplace_linalg_eigvalsh_cpu_float64`. Saw similar problems before.

Despite the param _being present_ in `out_wrapper`, the error was:

```
TypeError: meta__linalg_eigh() got an unexpected keyword argument 'eigenvalues'
```

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**